r/src/DataType.cpp

```diff
@@ -0,0 +1,212 @@
+
+#include "arrow_types.h"
+
+using namespace Rcpp;
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Int8__initialize() { return arrow::int8(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Int16__initialize() { return arrow::int16(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Int32__initialize() { return arrow::int32(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Int64__initialize() { return arrow::int64(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> UInt8__initialize() { return arrow::uint8(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> UInt16__initialize() { return arrow::uint16(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> UInt32__initialize() { return arrow::uint32(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> UInt64__initialize() { return arrow::uint64(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Float16__initialize() { return arrow::float16(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Float32__initialize() { return arrow::float32(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Float64__initialize() { return arrow::float64(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Boolean__initialize() { return arrow::boolean(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Utf8__initialize() { return arrow::utf8(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Date32__initialize() { return arrow::date32(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Date64__initialize() { return arrow::date64(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Null__initialize() { return arrow::null(); }
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Decimal128Type__initialize(int32_t precision, int32_t scale) {
+  return arrow::decimal(precision, scale);
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> FixedSizeBinary__initialize(int32_t byte_width) {
+  return arrow::fixed_size_binary(byte_width);
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Timestamp__initialize1(arrow::TimeUnit::type unit) {
+  return arrow::timestamp(unit);
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Timestamp__initialize2(arrow::TimeUnit::type unit, const std::string& timezone) {
+  return arrow::timestamp(unit, timezone);
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Time32__initialize(arrow::TimeUnit::type unit) {
+  return arrow::time32(unit);
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::DataType> Time64__initialize(arrow::TimeUnit::type unit) {
+  return arrow::time64(unit);
+}
+
+// [[Rcpp::export]]
+SEXP list__(SEXP x) {
+  if (Rf_inherits(x, "arrow::Field")) {
+    return wrap(arrow::list(Rcpp::as<std::shared_ptr<arrow::Field>>(x)));
+  }
+
+  if (Rf_inherits(x, "arrow::DataType")) {
+    return wrap(arrow::list(Rcpp::as<std::shared_ptr<arrow::DataType>>(x)));
+  }
+
+  stop("incompatible");
```

Comment:
Is there a generic "type error" that we can emit (more informative than this) when a type of SEXP is not supported?